### PR TITLE
Add initial support for opam 2.2 on Windows

### LINKED ISSUE #733: Native Windows Support
Fill out later!

### LINKED ISSUE #793: opam 2.2 support on Windows
The following code https://github.com/ocaml/setup-ocaml/blob/1903ece104ed71374b15d2bc25624ed60721c0ab/packages/setup-ocaml/src/constants.ts#L45-L46 disallows the use of the latest opam beta which is compatible with Windows.

Would it be possible to remove this conditional?

### PR DESCRIPTION
**Pull request by kit-ty-kate**
Closes #733
Closes #793

Joint work with @hhugo @smorimoto @dra27.

### CODE REVIEW

**Comment by smorimoto** [comment]
You can compile by running `yarn build`.

**Comment by smorimoto** [comment]
Could you rebase to the latest master?

**Comment by kit-ty-kate** [comment]
Note from @dra27: this should be part of `ocaml/setup-ocaml@v3`

**Comment by hhugo** [comment]
@kit-ty-kate, do you have some CI testing this GH action  ?

**Comment by kit-ty-kate** [comment]
@hhugo I do in https://github.com/kit-ty-kate/mirage-torrent/pull/2 but while the setup-ocaml succeeds, the builds fails while building dune:
```
#=== ERROR while compiling dune.3.15.2 ========================================#
# context     2.2.0~beta2 | win32/x86_64 | ocaml-base-compiler.4.14.2 | git+https://github.com/dra27/opam-repository.git#windows-initial
# path        D:\a\mirage-torrent\mirage-torrent\_opam\.opam-switch\build\dune.3.15.2
# command     D:\a\mirage-torrent\mirage-torrent\_opam\bin\ocaml.exe boot/bootstrap.ml -j 4
# exit-code   2
# env-file    D:\.opam\log\dune-8180-dff03d.env
# output-file D:\.opam\log\dune-8180-dff03d.out
### output ###
# ocamlc -output-complete-exe -w -24 -g -o .duneboot.exe -I boot unix.cma boot/libs.ml boot/duneboot.ml
# ** Cannot resolve symbols for D:/a/mirage-torrent/mirage-torrent/_opam/lib/ocaml\libcamlrun.a(startup_aux.b.o):
#  swscanf
# ** Cannot resolve symbols for descriptor object:
#  _vscwprintf
# File "boot/duneboot.ml", line 1:
# Error: Error while building custom runtime system
```

**Comment by hhugo** [comment]
I'm also testing this PR in https://github.com/hhugo/ocamlbuild/pull/7 and see the same error.

**Comment by hhugo** [comment]
The following can probably be removed as well
```
    // Works around the lack of https://github.com/ocaml/opam/pull/3882 when
    // adding ocaml/opam-repository on Windows. Can be removed when the action
    // switches to opam 2.2
    if (PLATFORM === "win32") {
      const autocrlf = await getExecOutput(
        "git",
        ["config", "--global", "core.autocrlf"],
        { ignoreReturnCode: true },
      );
      if (autocrlf.stdout.trim() !== "input") {
        if (autocrlf.exitCode === 0) {
          restore_autocrlf = autocrlf.stdout.trim();
        } else {
          restore_autocrlf = null; // Unset the value at the end
        }
      }
      await exec("git", ["config", "--global", "core.autocrlf", "input"]);
    }
```

File: .github/workflows/workflow.yml [hunk 1]
```diff
@@ -51,19 +51,11 @@ jobs:
           - ubuntu-latest
```
Comment (dra27) [inline .github/workflows/workflow.yml:51]:
```suggestion
          - ubuntu-latest
          - windows-latest
```

File: .github/workflows/workflow.yml [hunk 2]
```diff
@@ -51,19 +51,11 @@ jobs:
...
         ocaml-compiler:
           - "5.2"
-        allow-prerelease-opam:
-          - false
         include:
           - os: windows-latest
             ocaml-compiler: ocaml-variants.5.2.0+options,ocaml-option-mingw
```
Comment (dra27) [inline .github/workflows/workflow.yml:56]:
```suggestion
```

**Review by dra27 — COMMENTED** [review]
No need to special case WIndows in the test matrix any more 😎

**Review by smorimoto — APPROVED** [review]

**Comment by hhugo** [comment]
I'm testing this PR in https://github.com/ocsigen/js_of_ocaml/pull/1616, I still have issues on windows with git submodules but I suspect the issue is in opam not in setup-ocaml

File: packages/lint-doc/src/opam.ts [hunk 3]
```diff
@@ -3,18 +3,12 @@ import { exec } from "@actions/exec";
...
 export async function installOdoc() {
   await core.group("Install odoc", async () => {
-    await exec("opam", [
-      "depext",
-      "--install",
-      "conf-m4",
-      "dune",
-      "odoc>=1.5.0",
-    ]);
+    await exec("opam", ["install", "conf-m4", "dune", "odoc>=1.5.0"]);
```
Comment (kit-ty-kate) [inline packages/lint-doc/src/opam.ts:12]: m4 hasn't been needed in any of the core packages for over 3 years (the dependency was removed from ocamlfind in 2021)
```suggestion
    await exec("opam", ["install", "dune", "odoc>=1.5.0"]);
```

File: packages/setup-ocaml/src/opam.ts [hunk 4]
```diff
@@ -120,6 +120,22 @@ export async function setupOpam() {
...
+    path.join(OPAM_SWITCH_PREFIX, "lib", "ocaml"),
+  ].join(path.delimiter);
+  const OCAML_TOPLEVEL_PATH = path.join(OPAM_SWITCH_PREFIX, "lib", "toplevel");
+  const OPAM_SWITCH_BIN = path.join(OPAM_SWITCH_PREFIX, "bin");
+  core.exportVariable("OPAM_SWITCH_PREFIX", OPAM_SWITCH_PREFIX);
+  core.exportVariable("CAML_LD_LIBRARY_PATH", CAML_LD_LIBRARY_PATH);
+  core.exportVariable("OCAML_TOPLEVEL_PATH", OCAML_TOPLEVEL_PATH);
+  core.addPath(OPAM_SWITCH_BIN);
+}
+
```
Comment (dra27) [inline packages/setup-ocaml/src/opam.ts]: This doesn't work, I'm afraid - you either need to run `opam env` and parse its output, not guess the values.

FWIW, I think it's better for CI to have to be explicit about using `opam exec`

Comment (smorimoto) [reply packages/setup-ocaml/src/opam.ts]: Parsing the output sounds a bit overkill. The commit was dropped.

**Comment by hhugo** [comment]
When can we expect a release with this PR ?

File: packages/setup-ocaml/src/opam.ts (hunk 4 above)
Comment (smorimoto) [reply packages/setup-ocaml/src/opam.ts]: @dra27 For the record, could you explain why this approach doesn't work?

**Comment by smorimoto** [comment]
I'm planning to release as follows, but do you all have any objections? @dra27 @hhugo @kit-ty-kate

1. Merge this PR to `master`
2. Release as `v3.0.0-alpha` (maybe we shouldn't create the `v3` tag at this point)
3. Once we have tested widely and verified that there is no major problem, then release `v3.0.0`

**Comment by hhugo** [comment]
Sounds reasonable to me

**Comment by kit-ty-kate** [comment]
lgtm

**Comment by smorimoto** [comment]
https://github.com/ocaml/setup-ocaml/releases/tag/v3.0.0-alpha

**Pull request merged**